## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [189]:
import pandas as pd

df = pd.read_csv("Resources/client_dataset.csv")
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [190]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [191]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [192]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.shape
df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  
0            wall art     1096.80     762.71         7.50  105            1  
1                pens       24.95      15.09         1.49   21            0  
2  project management       13.52       7.86         1.68   39            6  
3                pens       36.42      24.85         1.23   29            3  
4                misc      195.10     108.17        46.43   20            1  
5              plants      172.84     154.44         2.79   78            0  
6    kitchen supplies      245.56     231.95         1.73   56            5  
7                pens       20.99      16.97         5.12   22            4  
8      paper products       83.86      59.00         0.27   49            3  
9             seating       66.88      54.15         2.47   78            1

In [193]:
# What three item categories had the most entries?
top_three_cat = df["category"].value_counts()
top_three_cat.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [194]:
# For the category with the most entries, which subcategory had the most entries?
top_category = df["category"].value_counts().index[0]
top_category = df[df["category"] == top_category]

top_subcategory = top_category["subcategory"].value_counts()
top_subcategory.head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [195]:
# Which five clients had the most entries in the data?
top_clients = df["client_id"].value_counts()
top_clients.head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [196]:
# Store the client ids of those top 5 clients in a list.
top_clients_list = top_clients.index.tolist()
top_clients_list

[33615,
 66037,
 46820,
 38378,
 24741,
 42481,
 71958,
 32500,
 49687,
 43607,
 84156,
 91091,
 62566,
 37715,
 41491,
 97876,
 80977,
 51869,
 84446,
 89110,
 29718,
 92201,
 78965,
 37542,
 21654,
 14935,
 95847,
 38473,
 48855,
 10520,
 29992,
 57574,
 49851,
 31715,
 16624,
 32597,
 10559,
 43361,
 79361,
 24007,
 35176,
 50202,
 79827,
 72892,
 18574,
 35794,
 16084,
 90851,
 49037,
 17699,
 37608,
 70745,
 38220,
 86106,
 57920,
 81150,
 34706,
 80857,
 10435,
 82075,
 21474,
 81576,
 49404,
 70029,
 65687,
 34507,
 55191,
 94675,
 33956,
 65748,
 58133,
 74559,
 32549,
 69528,
 33804,
 42550,
 41908,
 87612,
 31229,
 19405,
 85413,
 84344,
 71992,
 96954,
 28331,
 47078,
 82172,
 65819,
 79817,
 49058,
 86810,
 12286,
 34878,
 46709,
 73876,
 96736,
 62639,
 94136,
 46973,
 25119,
 61647,
 24485,
 25185,
 84432,
 44673,
 38192,
 77306,
 62992,
 19904,
 97124,
 26120,
 32240,
 10753,
 81012,
 47930,
 54350,
 56977,
 52678,
 73677,
 73362,
 53348,
 87985,
 26601,
 83686,
 22262,


In [197]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = top_clients_list[0]
total_units_top_client = df[df["client_id"] == top_client_id]["qty"].sum()
total_units_top_client

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [198]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["line_subtotal"] = df["unit_price"] * df["qty"]
df[["unit_price", "qty", "line_subtotal"]].head(2)


unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [199]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calc_shipping_price(x):
    if x > 50:
        return x * 7
    elif x <= 50:
        return x * 10
    
df["total_weight"] = df["unit_weight"] * df["qty"]
df["shipping_price"] = [calc_shipping_price(x) for x in df["total_weight"]]
df[["unit_price", "unit_weight", "qty", "total_weight", "shipping_price"]].head(3)


unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [200]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax_rate = 0.0925
df["line_price"] = ((df["line_subtotal"] + df["shipping_price"]) * (1 + sales_tax_rate)).round(2)
df[["line_subtotal", "shipping_price", "line_price"]].head(3)

line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [201]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["line_cost"] = (df["unit_cost"] * df["qty"]) + df["shipping_price"]

df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  total_weight  shipping_price  line_price  \
0            1      115164.00        787.50         5512.50   131839.08   
1            0         523.95         31.29          312.90      914.26   
2            6         527.28         65.52          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 23 columns]

In [202]:
# Create a column for the profit of each line using line cost and line price
df["line_profit"] = df["line_price"] - df["line_cost"]
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [203]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]

for order_id in order_ids:
    calculated_total = df[df["order_id"] == order_id]["line_price"].sum()
    print(f"Order {order_id} total: ${calculated_total:.2f}")

Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [204]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for client_id in top_clients_list:
    calculated_total = df[df["client_id"] == client_id]["line_price"].sum()
    print(f"{client_id}: ${calculated_total:.2f}")


33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02
42481: $7478305.34
71958: $14931663.18
32500: $8525901.60
49687: $25892348.48
43607: $12271034.25
84156: $19567017.42
91091: $15174095.50
62566: $15368088.58
37715: $6109455.57
41491: $7950554.43
97876: $8739116.99
80977: $16910899.72
51869: $7394156.24
84446: $5236275.08
89110: $13903411.93
29718: $8925031.45
92201: $12731862.32
78965: $655923670.05
37542: $15495183.37
21654: $6400545.38
14935: $13098525.94
95847: $6476224.43
38473: $183552571.81
48855: $12737794.07
10520: $6969902.97
29992: $5519097.12
57574: $35370314.95
49851: $7019318.87
31715: $8001697.01
16624: $5427092.19
32597: $8287254.01
10559: $7262527.81
43361: $5809959.68
79361: $4604756.83
24007: $5687328.41
35176: $9854461.78
50202: $9370675.40
79827: $12982285.55
72892: $4102931.88
18574: $36027948.56
35794: $9177409.23
16084: $9579881.75
90851: $5425432.50
49037: $9840221.13
17699: $8546617.74
37608: $7781781.78
70745: $5

In [205]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = pd.DataFrame([
    {
        "client_id": client_id,
        "qty": df[df["client_id"] == client_id]["qty"].sum(),
        "shipping_price": df[df["client_id"] == client_id]["shipping_price"].sum(),
        "line_price": df[df["client_id"] == client_id]["line_price"].sum(),
        "line_cost": df[df["client_id"] == client_id]["line_cost"].sum(),
        "line_profit": df[df["client_id"] == client_id]["line_profit"].sum()
    }
    for client_id in top_clients_list
])
summary_top_5_df = summary_df.head()

In [206]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ["shipping_price", "line_price", "line_cost", "line_profit"]

# Define a function that converts a dollar amount to millions.
def convertDollarsToMillions(amount):
    return amount / 1_000_000

# Apply the currency_format_millions function to only the money columns. 
for column in money_columns:
    summary_top_5_df[column] = summary_top_5_df[column].apply(convertDollarsToMillions)

# Rename the columns to reflect the change in the money format. 
reorg_df = summary_top_5_df.rename(columns={
    "client_id": "Client ID",
    "qty": "Units",
    "shipping_price": "Shipping (millions)",
    "line_price": "Total Revenue (millions)",
    "line_cost": "Total Cost (millions)",
    "line_profit": "Total Profit (millions)"
})

reorg_df.head()


C:\Users\T\AppData\Local\Temp\ipykernel_24336\3460404976.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  summary_top_5_df[column] = summary_top_5_df[column].apply(convertDollarsToMillions)


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [207]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_df = reorg_df.sort_values("Total Profit (millions)", ascending=False)
sorted_df.head()

# Based on the data analysis and manipulations performed, Client 24741 emerged as the most profitable, generating a total profit of approximately $36.58 million. 
# This client also demonstrated the highest revenue at around $82.27 million. 

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995